File: src/main/java/org/telran/lecture_14_hash/practice/SimpleHashTable.js
/**
 * Класс, представляющий хэш-таблицу.
 */
class HashTable {
    /**
     * Создает экземпляр хэш-таблицы.
     * @param {number} [size=10] - Размер хэш-таблицы.
     */
    constructor(size = 10) {
        this.table = new Array(size);
        this.size = size;
    }

    /**
     * Вычисляет хэш-код для заданного ключа.
     * @param {any} key - Ключ для вычисления хэш-кода.
     * @returns {number} - Хэш-код ключа.
     */
    hash(key) {
        if (typeof key !== 'number') {
            // Преобразуем строку в число (простой пример)
            let hash = 0;
            for (let i = 0; i < key.length; i++) {
                hash += key.charCodeAt(i);
            }
            key = hash;
        }
        const A = 0.6180339887; // Константа (золотое сечение)
        const fractionalPart = (key * A) % 1;
        return Math.floor(fractionalPart * this.size);
    }

    /**
     * Добавляет пару "ключ-значение" в хэш-таблицу.
     * @param {any} key - Ключ.
     * @param {any} value - Значение.
     */
    set(key, value) {
        let hash = this.hash(key);
        if (this.table[hash] === undefined) {
            this.table[hash] = [[key, value]];
        } else { // коллизия
            this.table[hash].push([key, value]);
        }
    }

    /**
     * Возвращает значение по заданному ключу.
     * @param {any} key - Ключ.
     * @returns {any|undefined} - Значение, связанное с ключом, или undefined, если ключ не найден.
     */
    get(key) {
        let hash = this.hash(key);
        let pairs = this.table[hash];
        for (let pair of pairs){
            if (pair[0] === key){
                return pair[1];
            }
        }
    }

    /**
     * Проверяет, есть ли ключ в хэш-таблице.
     * @param {any} key - Ключ.
     * @returns {boolean} - true, если ключ найден, иначе false.
     */
    has(key) {
        let hash = this.hash(key);
        let pairs = this.table[hash];
        for (let pair of pairs){
            if (pair[0] === key){
                return true;
            }
        }
        return false;
    }

    /**
     * Удаляет пару "ключ-значение" из хэш-таблицы.
     * @param {any} key - Ключ.
     */
    remove(key) {
        let hash = this.hash(key);
        let pairs = this.table[hash];
        for (let pair of pairs){
            if (pair[0] === key){
                this.removeSubarray(pairs, pair);
                return;
            }
        }
    }

    removeSubarray(arr, subarrayToRemove) {
        arr.forEach((item, index) => {
            if (Array.isArray(item) && JSON.stringify(item) === JSON.stringify(subarrayToRemove)) {
                arr.splice(index, 1);
            }
        });
    }

    equals(arr1, arr2){
        if (arr1.length !== arr2.length) {return false}
        for (let i = 0; i < arr1.length; i ++){
            if (arr1[i] !== arr2[i]) { return false}
        }
        return true;
    }

    /**
     * Выводит содержимое хэш-таблицы в консоль.
     */
    display() {
        for (let i = 0; i < this.size; i++) {
            if (this.table[i]) {
                console.log(i, ":", this.table[i]);
            }
        }
    }
}


// Пример использования
const myHashTable = new HashTable();
myHashTable.set("apple", "apple");
myHashTable.set("banana", "banana");
myHashTable.set("orange", "orange");
myHashTable.set("kivy", "kivy");
myHashTable.set("guava", "guava");
myHashTable.set("lemon", "lemon");
myHashTable.set("pineapple", "pineapple");
myHashTable.set("red_apple", "red_apple");
myHashTable.set("green_apple", "green_apple");
myHashTable.display();
console.log("--------------------------------")
myHashTable.remove("red_apple");
myHashTable.remove("guava");
myHashTable.display();
// console.log(myHashTable.get("pineapple"))
// console.log(myHashTable.get("red_apple"))
// console.log(myHashTable.has("orange"))
// console.log(myHashTable.has("nope"))
// 0: [["apple", "apple"], ["lemon", "lemon"]]

File: src/main/java/org/telran/lecture_14_hash/practice/SimpleHashTable.java
package org.telran.lecture_14_hash.practice;

import  java.util.LinkedList;

/**
 * Класс, представляющий хэш-таблицу.
 */
class SimpleHashTable {
    private Object[] table;
    private int size;

    /**
     * Создает экземпляр хэш-таблицы.
     * @param size Размер хэш-таблицы.
     */
    public SimpleHashTable(int size) {
        this.size = size;
        this.table = new Object[size];
    }

    /**
     * Создает экземпляр хэш-таблицы с размером по умолчанию (10).
     */
    public SimpleHashTable() {
        this(10);
    }

    /**
     * Вычисляет хэш-код для заданного ключа.
     * @param key Ключ для вычисления хэш-кода.
     * @return Хэш-код ключа.
     */
    public int hash(Object key) {
        int hash = 0;
        if (key instanceof String) {
            String strKey = (String) key;
            for (int i = 0; i < strKey.length(); i++) {
                hash += strKey.charAt(i);
            }
        } else if (key instanceof Integer) {
            hash = (Integer) key;
        } else {
            hash = key.hashCode();
        }
        return hash % size;
    }

    /**
     * Добавляет пару "ключ-значение" в хэш-таблицу.
     * @param key Ключ.
     * @param value Значение.
     */
    public void set(Object key, Object value) {
        int index = hash(key);
        if (table[index] == null) {
            table[index] = new LinkedList<>();
        }
        ((LinkedList<Pair>) table[index]).add(new Pair(key, value));
    }

    /**
     * Возвращает значение по заданному ключу.
     * @param key Ключ.
     * @return Значение, связанное с ключом, или null, если ключ не найден.
     */
    public Object get(Object key) {
        int index = hash(key);
        if (table[index] != null) {
            LinkedList<Pair> list = (LinkedList<Pair>) table[index];
            for (Pair pair : list) {
                if (pair.key.equals(key)) {
                    return pair.value;
                }
            }
        }
        return null;
    }

    /**
     * Проверяет, есть ли ключ в хэш-таблице.
     * @param key Ключ.
     * @return true, если ключ найден, иначе false.
     */
    public boolean has(Object key) {
        return get(key) != null;
    }

    /**
     * Удаляет пару "ключ-значение" из хэш-таблицы.
     * @param key Ключ.
     */
    public void remove(Object key) {
        int index = hash(key);
        if (table[index] != null) {
            LinkedList<Pair> list = (java.util.LinkedList<Pair>) table[index];
            list.removeIf(pair -> pair.key.equals(key));
            if (list.isEmpty()) {
                table[index] = null;
            }
        }
    }

    /**
     * Выводит содержимое хэш-таблицы в консоль.
     */
    public void display() {
        for (int i = 0; i < size; i++) {
            if (table[i] != null) {
                System.out.println(i + ": " + table[i]);
            }
        }
    }

    /**
     * Внутренний класс для хранения пары "ключ-значение".
     */
    private static class Pair {
        Object key;
        Object value;

        Pair(Object key, Object value) {
            this.key = key;
            this.value = value;
        }

        @Override
        public String toString() {
            return "{" + key + ": " + value + "}";
        }
    }

    public static void main(String[] args) {
        // Пример использования
        SimpleHashTable myHashTable = new SimpleHashTable();
        myHashTable.set("apple", 10);
        myHashTable.set("banana", 20);
        myHashTable.set("orange", 30);
        myHashTable.set("kivy", 35);
        myHashTable.set("guava", 3);
        myHashTable.set("lemon", 3);
        myHashTable.set("pineapple", 3);
        myHashTable.set("red_apple", 12);
        myHashTable.set("green_apple", 12);
        myHashTable.display();
    }
}